### ISSUE TITLE
Inconsistent behavior on default value to generically typed method

### ISSUE THREAD

**Issue by cmekik** [python/mypy#9793]
**Bug Report**

I get an error when I add a default value to a generically typed argument for a method in a generic class. But, when I do the same thing in a slightly roundabout way, the error is not raised.

**To Reproduce**

In the code below, the error occurs on the default `A` to the argument `constructor` of the `__init__` method for class `B`. 

The error message is:
Incompatible default for argument "constructor" (default has type "Type[A]", argument has type "Type[T]")

Class `B2`, on the other hand, does not raise an error, even though I think it has essentially the same effect.

```python
from typing import Hashable, Mapping, Generic, TypeVar, Type

T = TypeVar("T", bound="A")

class A:
    pass

class B(Generic[T]):

    mapping: Mapping[Hashable, T]
    constructor: Type[T]

    def __init__(
        self, 
        mapping: Mapping[Hashable, T], 
        constructor: Type[T] = A
    ) -> None:

        self.mapping = mapping
        self.constructor = constructor


class B2(Generic[T]):

    mapping: Mapping[Hashable, T]
    constructor: Type[T]

    def __init__(
        self, 
        mapping: Mapping[Hashable, T], 
        constructor: Type[T] = None
    ) -> None:

        self.mapping = mapping
        self.constructor = constructor if constructor is not None else A 
```

**Expected Behavior**

I expect `A` to be the default type for the type variable `T` while giving the user the option to subclass `A` if necessary. In my application, there are many possible variants of `A` and I would like to have one container class (represented by `B`/`B2`) that is aware of the specific mapping value type (which should not be changed after initialization). For instance, assuming `A2` is a subclass of `A`, I expect the following not to raise an error (it does not, ignoring the issue with the default value).

```python
a: A2 = B({1: A2}, A2).mapping[1]
``` 

**Actual Behavior**

I did not expect mypy to raise an error on the init method for class `B`. I also expected that if `B` raises an error so should `B2`, but it does not. 

**Your Environment**

I am using mypy as my linter in visual studio code.

- Mypy version used: 0.770
- Mypy command-line flags: none/default
- Mypy configuration options from `mypy.ini` (and other config files): default
- Python version used: 3.7
- Operating system and version: Ubuntu 20.04

**Comment by gvanrossum** [comment]
Instead of default argument values, can you try overloading different signatures of the constructor?

**Comment by hauntsaninja** [comment]
This is a duplicate of #3737, #8708, #8739, #9179, #9432 and others. First of those has the most discussion (see Guido's and Anders' comments in there). You don't get the error for B2, because `--no-implicit-optional` is not yet the default (alternatively use `--strict`).

**Comment by cmekik** [comment]
Thank you @gvanrossum and @hauntsaninja for the suggestions and links! I'm sorry I didn't find those earlier issues. I'll look more carefully next time. 

Overloading is indeed a good solution in my case. Here is a sample solution with some test cases.

```python
from typing import Hashable, Mapping, Generic, TypeVar, Type, overload


T = TypeVar("T", bound="A")


class A():
    pass


class A2(A):
    pass


class X():
    pass


class B3(Generic[T]):

    mapping: Mapping[Hashable, T]
    constructor: Type[T]

    @overload
    def __init__(self: B3[A], mapping: Mapping[Hashable, A]) -> None: 
        ...

    @overload
    def __init__(self: B3[T], constructor: Type[T]) -> None:
        ...

    @overload
    def __init__(
        self: B3[T], mapping: Mapping[Hashable, T], constructor: Type[T]
    ) -> None: 
        ...

    def __init__(self, mapping, constructor=A):

        self.mapping = mapping 
        self.constructor = constructor 


# Most typical use case, works as expected.
b3_1 = B3({}) 
a3_1: A = b3_1.mapping[1] # Will raise KeyError, but type check is fine.

# Next most typical use case, works as expected.
b3_2 = B3(constructor=A2)
a3_2: A2 = b3_2.mapping[1] # Will raise KeyError, but type check is fine.

# Some less typical cases, all as expected.

b3_3 = B3({1: A()}) 
a3_3: A = b3_3.mapping[1] # Type check is fine, as expected

b3_4 = B3({1: A2()}, A2)
a3_4: A2 = b3_4.mapping[1] # Type check is fine, as expected

# Expected error: 
# Argument "constructor" to "B3" has incompatible type "Type[X]"; expected "Type[A]"
# Correct b/c X is not a subclass of A.
b3_6 = B3(constructor=X) 

# Expected error: Value of type variable "T" of "B3" cannot be "X"
# Correct b/c X is not a subclass of A.
b3_5 = B3({1: X()}) 

# Some false negatives for my use case. Mapping values should be of the same 
# type. Not ideal, but not a deal breaker. Could be enforced at runtime with 
# isinstance checks if absolutely necessary. I'm not sure if there is a more 
# elegant workaround.

b3_8 = B3({1: A2()}) 
b3_7 = B3({1: A2()}, A)
```
